packages/cell-toolbar/src/celltoolbartracker.ts

```diff
@@ -0,0 +1,372 @@
+import { PanelLayout, Widget } from '@lumino/widgets';
+import { CellToolbarWidget } from './celltoolbarwidget';
+import { EXTENSION_ID, ICellMenuItem } from './tokens';
+
+// icon svg import statements
+import addAboveSvg from '../style/icons/addabove.svg';
+import addBelowSvg from '../style/icons/addbelow.svg';
+import deleteSvg from '../style/icons/delete.svg';
+import duplicateSvg from '../style/icons/duplicate.svg';
+import moveDownSvg from '../style/icons/movedown.svg';
+import moveUpSvg from '../style/icons/moveup.svg';
+
+const DEFAULT_LEFT_MENU: ICellMenuItem[] = [];
+
+/**
+ * Widget cell toolbar class
+ */
+const CELL_BAR_CLASS = 'jp-cell-bar';
+
+/**
+ * Class for a cell whose contents overlap with the cell toolbar
+ */
+const TOOLBAR_OVERLAP_CLASS = 'jp-toolbar-overlap';
+
+/**
+ * Icons for use in toolbar.
+ *
+ * These are copied from icon.ts, which is not part of the webpack bundle
+ * because nothing is imported from it.
+ */
+export const addAboveIcon = new LabIcon({
+  name: `${EXTENSION_ID}:add-above`,
+  svgstr: addAboveSvg
+});
+export const addBelowIcon = new LabIcon({
+  name: `${EXTENSION_ID}:add-below`,
+  svgstr: addBelowSvg
+});
+export const deleteIcon = new LabIcon({
+  name: `${EXTENSION_ID}:delete`,
+  svgstr: deleteSvg
+});
+export const duplicateIcon = new LabIcon({
+  name: `${EXTENSION_ID}:duplicate`,
+  svgstr: duplicateSvg
+});
+export const moveDownIcon = new LabIcon({
+  name: `${EXTENSION_ID}:move-down`,
+  svgstr: moveDownSvg
+});
+export const moveUpIcon = new LabIcon({
+  name: `${EXTENSION_ID}:move-up`,
+  svgstr: moveUpSvg
+});
+
+/**
+ * Watch a notebook so that a cell toolbar appears on the active cell
+ */
+export class CellToolbarTracker implements IDisposable {
+  constructor(
+    panel: NotebookPanel,
+    commands: CommandRegistry,
+    settings: ISettingRegistry.ISettings | null
+  ) {
+    this._commands = commands;
+    this._panel = panel;
+    this._settings = settings;
+    this._previousActiveCell = this._panel.content.activeCell;
+
+    if (this._settings) {
+      this._onSettingsChanged();
+      this._settings.changed.connect(this._onSettingsChanged, this);
+    }
+
+    const notebookModel = this._panel.context.model;
+    const cells = notebookModel.cells;
+    cells.changed.connect(this.updateConnectedCells, this);
+
+    // Only add the toolbar to the notebook's active cell (if any) once it has fully rendered and been revealed.
+    panel.revealed.then(() => this._onActiveCellChanged(panel.content));
+
+    // Handle subsequent changes of active cell.
+    panel.content.activeCellChanged.connect(this._onActiveCellChanged, this);
+  }
+
+  _onActiveCellChanged(notebook: Notebook): void {
+    const activeCell = notebook.activeCell;
+    if (!activeCell) {
+      return;
+    }
+
+    if (
+      this._previousActiveCell !== null &&
+      this._previousActiveCell !== undefined
+    ) {
+      this._removeToolbar(this._previousActiveCell.model);
+    }
```

Comment:
We should always remove the toolbar on the previous cell to be safe in case we change the logic that is possible to have no active cell.

I also simplify the test on the previous active cell.

```suggestion
    if (this._previousActiveCell) {
      this._removeToolbar(this._previousActiveCell.model);
    }
    
    const activeCell = notebook.activeCell;
    if (!activeCell) {
      return;
    }
```